Improve USB protections description (#43)
canonical/ubuntu-security-documentation: docs/security-features/platform-protections/devices.rst

Devices
#######

> This section documents security features that protect external devices.

Reviewer: Would it maybe be more accurate to say "...security features that protect **against** malicious external devices"?

Author: Haha yes that would make much more sense. Thanks for catching that!

Included below is an example of the content that may be included in the 
:file:`/etc/usbauth/usbauth.conf` configuration file:

.. code-block:: text
   :caption: /etc/usbauth/usbauth.conf

   # Allow a specific keyboard model.
   allow 046d:c52b

   # Also allow any device that identifies as a Mass Storage device.
   allow class 08

The above rules can be broken down into the following components:

* ``allow``: This is the action. It instructs the system to authorize any USB device that matches the criteria that follow.
* ``046d:c52b``: This is the matcher. It specifically targets the device's ``VendorID:ProductID`` pair. This rule does not distinguish between two identical device models; it allows any device of that exact model to connect.
* ``class 08``: This is an alternative type of matcher. The ``class`` keyword tells ``usbauth`` to look at the device's function rather than its manufacturer ID. The 08 is the value, which is the official `USB Class Code for Mass Storage devices <https://www.usb.org/defined-class-codes>`_. This single rule would permit any flash drive, external hard drive, or card reader to connect, regardless of its vendor or product ID.

> You can learn more about ``usbauth`` through its `official manpages <https://manpages.ubuntu.com/manpages/focal/man1/usbauth.1.html>`_.

Reviewer 2: Oh, man, I had personally linked to the non-release-specific link, which currently redirects to questing. @yhontyk, what should we standardize on?

``usbguard generate-policy > /etc/usbguard/rules.conf`` to create a baseline policy that 
allows all currently connected devices. From that point on, any new device will be blocked by
default.

When a new device is connected, the ``usbguard`` daemon logs the event, and a user can use 
``usbguard list-devices`` to see the blocked device. To permanently allow it, a user can use
``usbguard allow-device <id>`` and then append the new rule to their :file:`rules.conf` file 
to ensure it persists after a reboot. This makes ``usbguard`` a powerful and dynamic tool 
for managing USB security on a running system.

While managing ``usbguard`` from the command-line is effective, it can be cumbersome
on a desktop system. To improve usability, the ``usbguard-notifier`` service is
available. This tool monitors the ``usbguard`` daemon and provides a desktop
notification whenever a device is blocked. This notification pop-up allows a user to
immediately authorize the device and add a permanent rule for it, transforming the
experience from a manual, command-line process into an interactive one. It is the
successor to older tools like ``usbguard-applet-qt``.

> Included below is an example of the content that may be included in the :file:`rules.conf`
> file after running ``usbguard generate-policy > /etc/usbguard/rules.conf``:

Reviewer 3: ```suggestion
Here's an example of what you might add in the :file:`rules.conf`
file after running ``usbguard generate-policy > /etc/usbguard/rules.conf``:
```

.. code-block:: text
   :caption: /etc/usbguard/rules.conf

   allow id 046d:c52b name "Unifying Receiver" serial "4071-DE-AD-BE-EF" via-port "usb3-port2" with-interface { 03:01:01 03:01:02 }
   allow id 046d:082d name "HD Pro Webcam C920" serial "BADA55C0" via-port "usb3-port1" with-interface { 0e:01:00 0e:02:00 }
   allow id 1d6b:0002 name "Linux Foundation 2.0 root hub" serial "" via-port "" with-interface { 09:00:00 }

> 

Reviewer 3: I'd give the same general advice for this section as I gave above--if possible, I think it'd be better to give more general context, using the example file for the example variables. This depends on your naming conventions though, which actually look different here than they did in the previous section on usbauth? For example, up there you say that `allow` is the "action", but here you say it's the "target".

So, for example, the first one could be something like:

**Target** (`allow`): This is [definition of "target"]. The example uses `allow`, which means [XX]. Other possible targets are `block` and `reject`.

* ``id 046d:c52b``: This is the ``VendorID:ProductID`` pair. This is a primary attribute for matching.
* ``name "Unifying Receiver"``: The human-readable device name. This is for a user's reference and is ignored by the matching engine.
* ``serial "4071-DE-AD-BE-EF"``: The device's unique serial number. If present, this makes the rule extremely specific to a single physical device.
* ``via-port "usb3-port2"``: The physical port the device is connected to. This can be used to enforce that a device is only allowed in a specific port.
* ``with-interface { 03:01:01 03:01:02 }`` This is often the most critical attribute for security. It specifies the exact functions (e.g., keyboard, mouse, mass storage) the device is allowed to have. This is powerful for managing composite devices, which present multiple functions at once. For example, a programmable keyboard might also act as a mass storage device to store its configuration. A strict rule can allow the keyboard interface (03:01:01) while blocking the mass storage interface (08:06:50), greatly reducing the attack surface.

> By combining multiple attributes, especially the ``with-interface`` check, ``usbguard`` makes
> it significantly more difficult, but not outright impossible, for a malicious device to
> bypass the guardrails put in place by ``usbgaurd``, offering a substantial reduction in risk.

Reviewer 3: ```suggestion
By combining multiple attributes, especially the ``with-interface`` check, ``usbguard`` makes
it significantly more difficult for a malicious device to
bypass the guardrails put in place by ``usbguard``. This provides a substantial reduction in risk.
```
